_posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd

```diff
@@ -0,0 +1,681 @@
+trace_fn <- function(state, pkr) {
+  list(pkr$inner_results$is_accepted,
+       pkr$inner_results$accepted_results$step_size)
+}
+
+res <- hmc %>% mcmc_sample_chain(
+  num_results = n_steps,
+  num_burnin_steps = n_burnin,
+  current_state = initial_logits,
+  trace_fn = trace_fn
+)
+```
+
+When sampling is finished, we can access the samples as `res$all_states`:
+
+```{r}
+mcmc_trace <- res$all_states
+mcmc_trace
+```
+
+```
+Tensor("mcmc_sample_chain/trace_scan/TensorArrayStack/TensorArrayGatherV3:0",
+shape=(500, 4, 48), dtype=float32)
+```
+
+This is the shape of the samples for `l`, the 48 per-tank logits: 500 samples times 4 chains times 48 parameters.
+
+From these samples, we can compute effective sample size and $rhat$ (alias `mcmc_potential_scale_reduction`):
+
+```{r}
+# Tensor("Mean:0", shape=(48,), dtype=float32)
+ess <- mcmc_effective_sample_size(mcmc_trace) %>% tf$reduce_mean(axis = 0L)
+
+# Tensor("potential_scale_reduction/potential_scale_reduction_single_state/sub_1:0", shape=(48,), dtype=float32)
+rhat <- mcmc_potential_scale_reduction(mcmc_trace)
+```
+
+Whereas diagnostic information is available in `res$trace`:
+
+```{r}
+# Tensor("mcmc_sample_chain/trace_scan/TensorArrayStack_1/TensorArrayGatherV3:0",
+# shape=(500, 4), dtype=bool)
+is_accepted <- res$trace[[1]] 
+
+# Tensor("mcmc_sample_chain/trace_scan/TensorArrayStack_2/TensorArrayGatherV3:0",
+# shape=(500,), dtype=float32)
+step_size <- res$trace[[2]] 
+```
+
+After this quick outline, let's move on to the topic promised in the title: multi-level modeling, or partial pooling. This time, we'll also take a closer look at sampling results and diagnostic outputs.
+
+## Multi-level tadpoles ^[yep, it's a quote]
+
+
+The multi-level model -- or _varying intercepts model_, in this case: we'll get to _varying slopes_ in a later post -- adds a _hyperprior_ to the model. Instead of deciding on a mean and variance of the normal prior the logits are drawn from, we let the model learn means and variances for individual tanks.
+These per-tank means, while being priors for the binomial logits, are assumed to be normally distributed, and are themselves regularized by a normal prior for the mean and an exponential prior for the variance.
+
+For the Stan-savvy, here is the Stan formulation of this model.
+
+```{r}
+model{
+    vector[48] p;
+    sigma ~ exponential( 1 );
+    a_bar ~ normal( 0 , 1.5 );
+    a ~ normal( a_bar , sigma );
+    for ( i in 1:48 ) {
+        p[i] = a[tank[i]];
+        p[i] = inv_logit(p[i]);
+    }
+    S ~ binomial( N , p );
+}
+```
+
+
+And here it is with TFP:
+
+```{r}
+m2 <- tfd_joint_distribution_sequential(
+  list(
+    # a_bar, the prior for the mean of the normal distribution of per-tank logits
+    tfd_normal(loc = 0, scale = 1.5),
+    # sigma, the prior for the variance of the normal distribution of per-tank logits
+    tfd_exponential(rate = 1),
+    # normal distribution of per-tank logits
```

Comment:
I'd make a note here that this refers to `l` in the next one in the chain

Reply:
Oh I see you elaborate on this later, but still would be nice to follow what you did above for `a_bar` and `sigma`